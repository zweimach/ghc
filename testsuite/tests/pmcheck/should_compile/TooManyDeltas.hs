-- | The function here exploit matches of arity 2 that split the uncovered set
-- in two. Too many for -fmax-pmcheck-models=0!
-- As a result, these functions elicit the symptoms describe in the warnings
-- message, e.g.
--   - False positives on exhaustivity
--   - Fails to report redundant matches
--
-- We used to turn redundant into inaccessible clauses, but SG was unable to
-- produce a testcase. See the comment below.
module TooManyDeltas where

data T = A | B

-- | Reports that a clause for _ _ is missing.
f :: T -> T -> ()
f A A = ()

-- SG: As of July 2020, g doesn't reproduce anymore.
-- Because we treat constructor matches lazily and push data con match
-- strictness into a preceding bang guard, The single place that calls the
-- throttling function will not regress in laziness. Note that by throttling we
-- can only "forget" the x /~ K constraint from unc_this, not the preceding
-- x /~ âŠ¥ constraint.

-- | Reports that the third clause is inaccessible, when really it is
-- redundant.
g :: T -> T -> ()
g _ A = ()
g A A = () -- inaccessible, correctly flagged
g A A = () -- redundant, used to be inaccessible (see above).
g _ _ = () -- (this one is not about exhaustivity)

-- | Fails to report that the second clause is redundant.
h :: T -> T -> ()
h A A = () -- covered, emits no warning
h A A = () -- redundant, not covered!
h _ _ = () -- (this one is not about exhaustivity)
