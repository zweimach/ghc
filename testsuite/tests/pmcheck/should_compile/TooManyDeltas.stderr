
TooManyDeltas.hs:16:1: warning:
    Pattern match checker ran into -fmax-pmcheck-models=0 limit, so
      • Redundant clauses might not be reported at all
      • Redundant clauses might be reported as inaccessible
      • Patterns reported as unmatched might actually be matched
    Increase the limit or resolve the warnings to suppress this message.

TooManyDeltas.hs:16:1: warning: [-Wincomplete-patterns (in -Wextra)]
    Pattern match(es) are non-exhaustive
    In an equation for ‘f’:
        Patterns not matched:
            A A
            A B
            B A
            B B

TooManyDeltas.hs:28:1: warning:
    Pattern match checker ran into -fmax-pmcheck-models=0 limit, so
      • Redundant clauses might not be reported at all
      • Redundant clauses might be reported as inaccessible
      • Patterns reported as unmatched might actually be matched
    Increase the limit or resolve the warnings to suppress this message.

TooManyDeltas.hs:29:1: warning: [-Woverlapping-patterns (in -Wdefault)]
    Pattern match has inaccessible right hand side
    In an equation for ‘g’: g A A = ...

TooManyDeltas.hs:30:1: warning: [-Woverlapping-patterns (in -Wdefault)]
    Pattern match is redundant
    In an equation for ‘g’: g A A = ...
